.github/workflows/downstream.yml

```diff
@@ -0,0 +1,90 @@
+# Note: this file is inspired by the downstream testing facilities in the SciML ecosystem
+# x-ref: https://github.com/SciML/SciMLBase.jl/blob/ffe68aebedee5915190623cb08160d7ef1fbcce0/.github/workflows/Downstream.yml
+
+name: Downstream
+on:
+  push:
+    branches:
+      - main
+    paths-ignore:
+      - 'AUTHORS.md'
+      - 'CITATION.bib'
+      - 'CONTRIBUTING.md'
+      - 'LICENSE.md'
+      - 'NEWS.md'
+      - 'README.md'
+      - '.zenodo.json'
+      - '.github/workflows/benchmark.yml'
+      - '.github/workflows/CompatHelper.yml'
+      - '.github/workflows/TagBot.yml'
+      - 'benchmark/**'
+      # - 'docs/**'
+      - 'utils/**'
+  pull_request:
+    paths-ignore:
+      - 'AUTHORS.md'
+      - 'CITATION.bib'
+      - 'CONTRIBUTING.md'
+      - 'LICENSE.md'
+      - 'NEWS.md'
+      - 'README.md'
+      - '.zenodo.json'
+      - '.github/workflows/benchmark.yml'
+      - '.github/workflows/CompatHelper.yml'
+      - '.github/workflows/TagBot.yml'
+      - 'benchmark/**'
+      # - 'docs/**'
+      - 'utils/**'
+  workflow_dispatch:
+
+# Cancel redundant CI tests automatically
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  test:
+    if: "!contains(github.event.head_commit.message, 'skip ci')"
+    # We could also include the Julia version as in
+    # name: ${{ matrix.trixi_test }} - ${{ matrix.os }} - Julia ${{ matrix.version }} - ${{ matrix.arch }} - ${{ github.event_name }}
+    # to be more specific. However, that requires us updating the required CI tests whenever we update Julia.
+    name: ${{ matrix.package }} - ${{ matrix.os }} - ${{ matrix.arch }} - ${{ github.event_name }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        version:
+          - '1.9'
+        os:
+          - ubuntu-latest
+        arch:
+          - x64
+        package:
+          - TrixiShallowWater.jl
+    steps:
+      - uses: actions/checkout@v4
+      - uses: julia-actions/setup-julia@v1
+        with:
+          version: ${{ matrix.version }}
+          arch: ${{ matrix.arch }}
+      - run: julia -e 'using InteractiveUtils; versioninfo(verbose=true)'
+      - uses: julia-actions/cache@v1
+      - uses: julia-actions/julia-buildpkg@v1
+      - name: Retrieve downstream package
+        uses: actions/checkout@v4
+        with:
+          repository: trixi-framework/${{ matrix.package }}
```

Comment:
This uses the current `main` branch of the downstream package, doesn't it? Is this what we want to do - or do we want to use the latest stable release of the downstream package? (This is an honest question)

Reply:
Good point! I thought about it a little, and I think the fundamental question to ask is when we want to be notified about potential issues/incompatibilities.

First, I think downstream packages such as TrixiSWE should only be required to test against releases (like we do with other upstream packages). It is unreasonable (and not standard practice) to require them to check against our `main`.

If we test the `main` branch of the downstream package, there are imho at least the following arguments to weigh:
* Pro:
  * If the downstream package uses additional parts of Trixi.jl in `main` that weren't used in a previous downstream release, and we break them (accidentally) in Trixi.jl, we will learn about it immediately. Specifically, it allows us to correct the issue before the downstream package potentially creates a new release that is broken (because their previous tests still worked).
* Con:
  * If the downstream package's `main` fails the tests for reasons unrelated to the upstream changes in Trixi.jl, we get false negatives (or false positives? 🤔 🤷‍♂️).

On the other hand, if we only test against released downstream versions, we would (maybe) add an incentive to create frequent releases in downstream packages (as we do for Trixi.jl anyways).

Since I do not have experience with either option, my suggestion would be to pick one (probably the currently implemented one) and then gather some experience with it.

Reply:
Sounds good to me. Could you please add a comment explaining this briefly to the CI job file?

Reply:
6d1ee667c